File: tests/e2e/pages/login_page.py
"""
Login Page Object
"""
from .base_page import BasePage
from ..config.test_config import TestConfig


class LoginPage(BasePage):
    """Login page interactions"""
    
    def __init__(self, page):
        super().__init__(page)
        self.url = TestConfig.FRONTEND_URL
        
    # Selectors
    LOGIN_BUTTON = '[data-testid="login-button"]'
    LOGOUT_BUTTON = '[data-testid="logout-button"]'
    USER_MENU = '[data-testid="user-menu"]'
    USERNAME_INPUT = '#username'
    PASSWORD_INPUT = '#password'
    KEYCLOAK_LOGIN_BUTTON = '#kc-login'
    ERROR_MESSAGE = '.alert-error'
    
    async def navigate(self):
        """Navigate to application"""
        await super().navigate(self.url)
        
    async def login(self, username: str, password: str):
        """Complete login flow"""
        # Click login button
        await self.click_element(self.LOGIN_BUTTON)
        
        # Wait for Keycloak redirect
        await self.page.wait_for_url('**/auth/realms/**', timeout=10000)
        
        # Fill credentials
        await self.fill_input(self.USERNAME_INPUT, username)
        await self.fill_input(self.PASSWORD_INPUT, password)
        
        # Submit form
        await self.click_element(self.KEYCLOAK_LOGIN_BUTTON)
        
        # Wait for redirect back to application
        await self.page.wait_for_url(self.url + '/**', timeout=15000)
        
        # Verify login success
        await self.wait_for_element(self.USER_MENU)
    
    async def logout(self):
        """Logout from application"""
        # Click user menu
        await self.click_element(self.USER_MENU)
        
        # Click logout
        await self.click_element(self.LOGOUT_BUTTON)
        
        # Wait for redirect
        await self.page.wait_for_url(self.url, timeout=10000)
        
        # Verify logout
        await self.wait_for_element(self.LOGIN_BUTTON)
        
    async def is_logged_in(self) -> bool:
        """Check if user is logged in"""
        return await self.is_visible(self.USER_MENU)
        
    async def get_username(self) -> str:
        """Get logged in username"""
        if await self.is_logged_in():
            return await self.get_text(self.USER_MENU)
        return None
        
    async def get_error_message(self) -> str:
        """Get login error message"""
        if await self.is_visible(self.ERROR_MESSAGE):
            return await self.get_text(self.ERROR_MESSAGE)
        return None

File: frontend/e2e/pages/login.page.ts
import { Page } from '@playwright/test';
import { BasePage } from './base.page';

export class LoginPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  // Locators
  get loginButton() {
    return this.getByTestId('login-button');
  }

  get logoutButton() {
    return this.getByTestId('logout-button');
  }

  get userMenu() {
    return this.getByTestId('user-menu');
  }

  get usernameInput() {
    return this.page.locator('#username');
  }

  get passwordInput() {
    return this.page.locator('#password');
  }

  get keycloakLoginButton() {
    return this.page.locator('#kc-login');
  }

  get errorMessage() {
    return this.page.locator('.alert-error, .kc-feedback-text');
  }

  // Actions
  async clickLogin() {
    await this.loginButton.click();
  }

  async performLogin(username: string, password: string) {
    // Click login button to redirect to Keycloak
    await this.clickLogin();
    
    // Wait for Keycloak page
    await this.page.waitForURL('**/auth/realms/**');
    
    // Fill credentials
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    
    // Submit
    await this.keycloakLoginButton.click();
    
    // Wait for redirect back
    await this.page.waitForURL(/http:\/\/localhost:\d+/);
  }

  async logout() {
    await this.userMenu.click();
    await this.logoutButton.click();
    await this.page.waitForURL(/http:\/\/localhost:\d+/);
  }

  async isLoggedIn(): Promise<boolean> {
    return await this.userMenu.isVisible();
  }

  async getUsername(): Promise<string | null> {
    if (await this.isLoggedIn()) {
      return await this.userMenu.textContent();
    }
    return null;
  }

  async getErrorMessage(): Promise<string | null> {
    if (await this.errorMessage.isVisible()) {
      return await this.errorMessage.textContent();
    }
    return null;
  }
}
